lib/archethic_web/controllers/api/web_hosting_controller.ex

```diff
@@ -57,41 +64,153 @@ defmodule ArchethicWeb.API.WebHostingController do
-           get_cache(cache_headers, last_address, url_path),
-         {:ok, file_content, encodage} <- get_file_content(file, cached?, url_path) do
-      {:ok, file_content, encodage, mime_type, cached?, etag}
+         {:ok, transaction} <- Archethic.get_last_transaction(address) do
+      with {:ok, json_content} <- Jason.decode(transaction.data.content),
+           {:ok, file, mime_type} <- get_file(json_content, url_path),
+           {cached?, etag} <- get_cache(cache_headers, transaction.address, url_path),
+           {:ok, file_content, encodage} <- get_file_content(file, cached?, url_path) do
+        {:ok, file_content, encodage, mime_type, cached?, etag}
+      else
+        :encodage_error ->
+          {:error, :invalid_encodage}
+
+        :file_error ->
+          {:error, :file_not_found}
+
+        {:error, %Jason.DecodeError{}} ->
+          {:error, :invalid_content}
+
+        {:error, :file_not_found} ->
+          {:error, :file_not_found}
+
+        {:error, :malformed} ->
+          # malformed file will return 404 as described in test "should return Cannot find file content"
+          {:error, :file_not_found}
+
+        {:error, :is_a_directory} ->
+          # return the transaction so the dir_listing function do not need to do the I/O
+          {:error, {:is_a_directory, transaction}}
+      end
     else
       er when er in [:error, false] ->
         {:error, :invalid_address}
 
-      {:error, %Jason.DecodeError{}} ->
-        {:error, :invalid_content}
-
       {:error, reason} when reason in [:transaction_not_exists, :transaction_invalid] ->
         {:error, :website_not_found}
 
-      {:file_not_found, _url} ->
-        {:error, :not_found}
+      error ->
+        error
+    end
+  end
+
+  @spec dir_listing(
+          request_path :: String.t(),
+          params :: map(),
+          transaction :: Transaction.t(),
+          cached_headers :: list()
+        ) ::
+          {:ok, listing_html :: binary() | nil, encodage :: nil | binary(), mime_type :: binary(),
+           cached? :: boolean(), etag :: binary()}
+  def dir_listing(request_path, params, transaction, cache_headers) do
+    url_path = Map.get(params, "url_path", [])
+
+    with {:ok, json_content} <- Jason.decode(transaction.data.content),
+         {cached?, etag} <- get_cache(cache_headers, transaction.address, url_path) do
+      json_content_subset =
+        case url_path do
+          [] ->
+            json_content
+
+          _ ->
+            {:ok, subset} = Pathex.view(json_content, get_json_path(url_path))
+            subset
+        end
+
+      {current_working_dir, parent_dir_href} =
+        case url_path do
+          [] ->
+            {"/", nil}
+
+          _ ->
+            {
+              Path.join(["/" | url_path]),
+              %{href: request_path |> Path.join("..") |> Path.expand()}
+            }
+        end
+
+      assigns =
+        Map.keys(json_content_subset)
+        |> Enum.map(fn key ->
```

Comment:
Here you can directly use `Enum.map` on the map and directly access to the value like
```
json_content_subset
|> Enum.map(fn
  {key, %{"address" => address}} -> {:file, key, address}
  {key, _} -> {:dir, key}
end)
```

This will simplify as we get the key and the value in the same instruction